drivers/segger/serial_rtt.c

```diff
@@ -0,0 +1,368 @@
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <assert.h>
+#include <debug.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include <nuttx/kmalloc.h>
+#include <nuttx/kthread.h>
+#include <nuttx/list.h>
+#include <nuttx/serial/serial.h>
+
+#include <SEGGER_RTT.h>
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+struct rtt_dev_s
+{
+  int channel;
+  FAR char *txbuf;
+  FAR char *rxbuf;
+  struct list_node node;
+  struct uart_dev_s uart;
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static int serial_rtt_setup(FAR struct uart_dev_s *dev);
+static void serial_rtt_shutdown(FAR struct uart_dev_s *dev);
+static int serial_rtt_attach(FAR struct uart_dev_s *dev);
+static void serial_rtt_detach(FAR struct uart_dev_s *dev);
+static int serial_rtt_ioctl(FAR struct file *filep, int cmd,
+                            unsigned long arg);
+static int serial_rtt_receive(FAR struct uart_dev_s *dev,
+                              unsigned int *status);
+static void serial_rtt_rxint(FAR struct uart_dev_s *dev, bool enable);
+static bool serial_rtt_rxavailable(FAR struct uart_dev_s *dev);
+static void serial_rtt_send(FAR struct uart_dev_s *dev, int ch);
+static void serial_rtt_txint(FAR struct uart_dev_s *dev, bool enable);
+static bool serial_rtt_txready(FAR struct uart_dev_s *dev);
+static bool serial_rtt_txempty(FAR struct uart_dev_s *dev);
+static void serial_rtt_dmasend(FAR struct uart_dev_s *dev);
+static void serial_rtt_dmareceive(FAR struct uart_dev_s *dev);
+static void serial_rtt_dmarxfree(FAR struct uart_dev_s *dev);
+static void serial_rtt_dmatxavail(FAR struct uart_dev_s *dev);
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+struct list_node g_rtt_dev_list = LIST_INITIAL_VALUE(g_rtt_dev_list);
+
+static const struct uart_ops_s g_serial_rtt_ops =
+{
+  .setup = serial_rtt_setup,
```

Comment:
Can we use C89 compatible struct init?

Reply:
This is implemented with reference to uart_rpmsg.c
What do you mean by C89 compatibility?

Reply:
C89 can't use the designated initializer:
https://stackoverflow.com/questions/47202557/what-is-a-designated-initializer-in-c

Reply:
done~